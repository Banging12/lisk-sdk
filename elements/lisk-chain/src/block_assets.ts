/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { blockAssetSchema } from './schema';

export interface BlockAsset {
	moduleID: number;
	data: Buffer;
}

export class BlockAssets {
	private readonly _assets: BlockAsset[] = [];

	public constructor(assets: BlockAsset[] = []) {
		this._assets = assets;
	}

	public static fromBytes(values: ReadonlyArray<Buffer>): BlockAssets {
		const assets = values.map(val => codec.decode<BlockAsset>(blockAssetSchema, val));
		const blockAssets = new BlockAssets(assets);
		return blockAssets;
	}

	public static fromJSON(values: Record<string, unknown>[]): BlockAssets {
		const assets = values.map(val => codec.fromJSON<BlockAsset>(blockAssetSchema, val));
		return new BlockAssets(assets);
	}

	public getBytes(): Buffer[] {
		return this._assets.map(asset => codec.encode(blockAssetSchema, asset));
	}

	public getAsset(moduleID: number): Buffer | undefined {
		return this._assets.find(a => a.moduleID === moduleID)?.data;
	}

	public setAsset(moduleID: number, value: Buffer): void {
		const asset = this.getAsset(moduleID);
		if (asset) {
			throw new Error(`Module asset for "${moduleID}" is already set.`);
		}

		this._assets.push({ moduleID, data: value });
	}

	public sort(): void {
		this._assets.sort((a1, a2) => a1.moduleID - a2.moduleID);
	}
}
