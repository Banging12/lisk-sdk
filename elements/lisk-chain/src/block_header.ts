/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { objects } from '@liskhq/lisk-utils';
import { signDataWithPrivateKey, hash, verifyData } from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import { validator, LiskValidationError } from '@liskhq/lisk-validator';
import { TAG_BLOCK_HEADER } from './constants';
import { blockHeaderSchema, blockHeaderSchemaWithId, signingBlockHeaderSchema } from './schema';

export interface BlockHeaderAsset {
	moduleID: number;
	data: Buffer;
}

export interface BlockHeaderAttrs {
	readonly version: number;
	readonly height: number;
	readonly generatorAddress: Buffer;
	readonly previousBlockID: Buffer;
	readonly timestamp: number;
	readonly stateRoot?: Buffer;
	readonly transactionRoot?: Buffer;
	readonly assets: ReadonlyArray<BlockHeaderAsset>;
	signature?: Buffer;
	id?: Buffer;
}

export interface BlockHeaderJSON {
	readonly version: number;
	readonly timestamp: number;
	readonly height: number;
	readonly generatorAddress: string;
	readonly previousBlockID: string;
	readonly stateRoot: string;
	readonly transactionRoot: string;
	readonly assets: ReadonlyArray<BlockHeaderAsset>;
	readonly signature: string;
	readonly id: string;
}

export class BlockHeader {
	public readonly version: number;
	public readonly height: number;
	public readonly generatorAddress: Buffer;
	public readonly previousBlockID: Buffer;
	public readonly timestamp: number;
	private _stateRoot?: Buffer;
	private _transactionRoot?: Buffer;
	private readonly _assets: BlockHeaderAsset[];
	private _signature?: Buffer;
	private _id?: Buffer;

	public constructor({
		version,
		timestamp,
		height,
		generatorAddress,
		previousBlockID,
		stateRoot,
		transactionRoot,
		signature,
		id,
		assets,
	}: BlockHeaderAttrs) {
		this.version = version;
		this.height = height;
		this.generatorAddress = generatorAddress;
		this.previousBlockID = previousBlockID;
		this.timestamp = timestamp;
		this._stateRoot = stateRoot;
		this._transactionRoot = transactionRoot;
		this._assets = objects.cloneDeep<BlockHeaderAsset[]>([...assets]);

		this._signature = signature;
		this._id = id;
	}

	public static fromBytes(value: Buffer): BlockHeader {
		return new BlockHeader(codec.decode<BlockHeaderAttrs>(blockHeaderSchema, value));
	}

	public static fromJSON(value: Record<string, unknown>): BlockHeader {
		return new BlockHeader(codec.fromJSON<BlockHeaderAttrs>(blockHeaderSchema, value));
	}

	public get stateRoot() {
		return this._stateRoot;
	}

	public set stateRoot(val) {
		this._stateRoot = val;
		this._resetComputedValues();
	}

	public get transactionRoot() {
		return this._transactionRoot;
	}

	public set transactionRoot(val) {
		this._transactionRoot = val;
		this._resetComputedValues();
	}

	public getBytes(): Buffer {
		return codec.encode(blockHeaderSchema, this._getBlockHeaderProps());
	}

	public toJSON(): BlockHeaderJSON {
		return codec.toJSON(blockHeaderSchemaWithId, this._getAllProps());
	}

	public toObject(): BlockHeaderAttrs {
		return this._getAllProps();
	}

	public validate(): void {
		const errors = validator.validate(blockHeaderSchema, this._getBlockHeaderProps());
		if (errors.length) {
			throw new LiskValidationError(errors);
		}
		if (this.previousBlockID.length === 0) {
			throw new Error('Previous block id must not be empty.');
		}
	}

	public validateSignature(publicKey: Buffer, networkIdentifier: Buffer): void {
		const signingBytes = this.getSigningBytes();
		const verified = verifyData(
			TAG_BLOCK_HEADER,
			networkIdentifier,
			signingBytes,
			this.signature,
			publicKey,
		);

		if (!verified) {
			throw new Error('Invalid block signature.');
		}
	}

	public getSigningBytes(): Buffer {
		const blockHeaderBytes = codec.encode(signingBlockHeaderSchema, this._getSigningProps());

		return blockHeaderBytes;
	}

	public sign(networkIdentifier: Buffer, privateKey: Buffer): void {
		this._signature = signDataWithPrivateKey(
			TAG_BLOCK_HEADER,
			networkIdentifier,
			this.getSigningBytes(),
			privateKey,
		);
	}

	public getAsset(moduleID: number): Buffer | undefined {
		return this._assets.find(a => a.moduleID === moduleID)?.data;
	}

	public setAsset(moduleID: number, value: Buffer): void {
		const asset = this.getAsset(moduleID);
		if (asset) {
			throw new Error(`Module asset for "${moduleID}" is already set.`);
		}

		this._assets.push({ moduleID, data: value });
		this._resetComputedValues();
	}

	public get signature(): Buffer {
		if (!this._signature) {
			throw new Error('Block header is not signed.');
		}

		return this._signature;
	}

	public get id(): Buffer {
		if (!this._id && !this._signature) {
			throw new Error('Can not generate the id for unsigned block header.');
		}

		if (!this._id) {
			const blockHeaderBytes = codec.encode(blockHeaderSchema, {
				...this._getSigningProps(),
				signature: this._signature,
			});

			this._id = hash(blockHeaderBytes);
		}

		return this._id;
	}

	private _resetComputedValues() {
		this._id = undefined;
		this._signature = undefined;
	}

	private _getSigningProps() {
		return {
			version: this.version,
			timestamp: this.timestamp,
			height: this.height,
			previousBlockID: this.previousBlockID,
			stateRoot: this.stateRoot,
			transactionRoot: this.transactionRoot,
			generatorAddress: this.generatorAddress,
			assets: this._assets,
		};
	}

	private _getBlockHeaderProps() {
		return {
			...this._getSigningProps(),
			signature: this.signature,
		};
	}

	private _getAllProps() {
		return {
			...this._getBlockHeaderProps(),
			id: this.id,
		};
	}
}
