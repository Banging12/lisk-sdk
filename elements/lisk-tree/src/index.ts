/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { verifyProof } from './merkle_tree/verify_proof';
import { calculateRootFromUpdateData } from './merkle_tree/calculate';
import { MerkleTree } from './merkle_tree/merkle_tree';
import { calculateMerkleRoot, calculateMerkleRootWithLeaves } from './merkle_tree/utils';

export const regularMerkleTree = {
	verifyProof,
	calculateRootFromUpdateData,
	calculateMerkleRoot,
	calculateMerkleRootWithLeaves,
	MerkleTree,
};

export { MerkleTree };
