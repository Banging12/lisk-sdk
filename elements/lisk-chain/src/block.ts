/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { regularMerkleTree } from '@liskhq/lisk-tree';
import { BlockAssets } from './block_assets';
import { BlockHeader } from './block_header';
import { MAX_ASSET_DATA_SIZE_BYTES } from './constants';
import { blockSchema } from './schema';
import { Transaction } from './transaction';

interface BlockAttrs {
	header: Buffer;
	payload: Buffer[];
	assets: Buffer[];
}

export class Block {
	// eslint-disable-next-line no-useless-constructor
	public constructor(
		public readonly header: BlockHeader,
		public readonly payload: Transaction[],
		public readonly assets: BlockAssets,
	) {
		// No body necessary
	}

	public static fromBytes(value: Buffer): Block {
		const { header, payload, assets } = codec.decode<BlockAttrs>(blockSchema, value);

		return new Block(
			BlockHeader.fromBytes(header),
			payload.map(v => Transaction.fromBytes(v)),
			BlockAssets.fromBytes(assets),
		);
	}

	public static fromJSON(value: Record<string, unknown>): Block {
		const { header, payload, assets } = value;
		if (typeof header !== 'object') {
			throw new Error('Invalid block format. header must be an object.');
		}
		if (!Array.isArray(payload)) {
			throw new Error('Invalid block format. payload must be an array.');
		}
		if (!Array.isArray(assets)) {
			throw new Error('Invalid block format. assets must be an array.');
		}

		return new Block(
			BlockHeader.fromJSON(value.header as Record<string, unknown>),
			payload.map(v => Transaction.fromBytes(v)),
			BlockAssets.fromJSON(assets),
		);
	}

	public getBytes(): Buffer {
		return codec.encode(blockSchema, {
			header: this.header.getBytes(),
			payload: this.payload.map(p => p.getBytes()),
			assets: this.assets.getBytes(),
		});
	}

	public validate(): void {
		this.header.validate();
		for (const tx of this.payload) {
			tx.validate();
		}

		const assets = this.assets.getAll();
		let last = assets[0];
		let i = 0;
		for (const asset of assets) {
			// Data size of each module should not be greater than max asset data size
			if (asset.data.byteLength > MAX_ASSET_DATA_SIZE_BYTES) {
				throw new Error(
					`Module with ID ${asset.moduleID} has data size more than ${MAX_ASSET_DATA_SIZE_BYTES} bytes.`,
				);
			}
			if (last.moduleID > asset.moduleID) {
				throw new Error('Assets are not sorted in the increasing values of moduleID.');
			}
			// Check for duplicates
			if (i > 0 && asset.moduleID === last.moduleID) {
				throw new Error(`Module with ID ${assets[i].moduleID} has duplicate entries.`);
			}
			i += 1;
			last = asset;
		}

		if (
			!this.header.transactionRoot?.equals(
				regularMerkleTree.calculateMerkleRootWithLeaves(this.payload.map(tx => tx.id)),
			)
		) {
			throw new Error('Block header transaction root is invalid.');
		}

		if (
			!this.header.assetsRoot?.equals(
				regularMerkleTree.calculateMerkleRootWithLeaves(this.assets.getBytes()),
			)
		) {
			throw new Error('Block header asset root is invalid.');
		}
	}
}
