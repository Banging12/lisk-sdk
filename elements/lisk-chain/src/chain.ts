/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { KVStore, NotFoundError } from '@liskhq/lisk-db';
import * as createDebug from 'debug';
import { MerkleTree } from '@liskhq/lisk-tree';
import {
	DEFAULT_MAX_BLOCK_HEADER_CACHE,
	DEFAULT_MIN_BLOCK_HEADER_CACHE,
	GENESIS_BLOCK_VERSION,
} from './constants';
import { DataAccess } from './data_access';
import { StateStore } from './state_store';
import { validateGenesisBlock } from './validate';
import {
	blockSchema,
	signingBlockHeaderSchema,
	blockHeaderSchema,
	stateDiffSchema,
} from './schema';
import { Block } from './block';
import { BlockHeader } from './block_header';

interface ChainConstructor {
	// Constants
	readonly maxPayloadLength: number;
	readonly minBlockHeaderCache?: number;
	readonly maxBlockHeaderCache?: number;
}

interface ChainInitArgs {
	readonly db: KVStore;
	readonly networkIdentifier: Buffer;
	readonly genesisBlock: Block;
}

const debug = createDebug('lisk:chain');

export class Chain {
	public dataAccess!: DataAccess;
	public readonly constants: {
		readonly maxPayloadLength: number;
		readonly minBlockHeaderCache: number;
		readonly maxBlockHeaderCache: number;
	};

	private _lastBlock?: Block;
	private _finalizedHeight?: number;
	private _networkIdentifier!: Buffer;

	public constructor({
		// Constants
		maxPayloadLength,
		minBlockHeaderCache = DEFAULT_MIN_BLOCK_HEADER_CACHE,
		maxBlockHeaderCache = DEFAULT_MAX_BLOCK_HEADER_CACHE,
	}: ChainConstructor) {
		// Register codec schema
		codec.addSchema(blockSchema);
		codec.addSchema(blockHeaderSchema);
		codec.addSchema(signingBlockHeaderSchema);
		codec.addSchema(stateDiffSchema);

		this.constants = {
			maxPayloadLength,
			maxBlockHeaderCache,
			minBlockHeaderCache,
		};
	}

	public get genesisHeight(): number {
		return this._genesisHeight;
	}

	public get lastBlock(): Block {
		if (!this._lastBlock) {
			throw new Error('Chain has not been initialized');
		}
		return this._lastBlock;
	}

	public get finalizedHeight(): number {
		if (!this._finalizedHeight) {
			throw new Error('Chain has not been initialized');
		}
		return this._finalizedHeight;
	}

	public get networkIdentifier(): Buffer {
		return this._networkIdentifier;
	}

	public init(args: ChainInitArgs): void {
		this._networkIdentifier = args.networkIdentifier;
		this.dataAccess = new DataAccess({
			db: args.db,
			minBlockHeaderCache: this.constants.minBlockHeaderCache,
			maxBlockHeaderCache: this.constants.maxBlockHeaderCache,
		});
	}

	public async loadLastBlocks(genesisBlock: Block): Promise<void> {
		let storageLastBlock: Block;
		try {
			storageLastBlock = await this.dataAccess.getLastBlock();
		} catch (error) {
			throw new Error('Failed to load last block');
		}

		if (storageLastBlock.header.height !== genesisBlock.header.height) {
			await this._cacheBlockHeaders(storageLastBlock);
		}
		this._lastBlock = storageLastBlock;
		this._finalizedHeight = await this.dataAccess.getFinalizedHeight();
	}

	public resetBlockHeaderCache(): void {
		this.dataAccess.resetBlockHeaderCache();
	}

	public async genesisBlockExist(genesisBlock: Block): Promise<boolean> {
		let matchingGenesisBlock: BlockHeader | undefined;
		try {
			matchingGenesisBlock = await this.dataAccess.getBlockHeaderByID(genesisBlock.header.id);
		} catch (error) {
			if (!(error instanceof NotFoundError)) {
				throw error;
			}
		}
		let lastBlockHeader: BlockHeader | undefined;
		try {
			lastBlockHeader = await this.dataAccess.getLastBlockHeader();
		} catch (error) {
			if (!(error instanceof NotFoundError)) {
				throw error;
			}
		}
		if (lastBlockHeader && !matchingGenesisBlock) {
			throw new Error('Genesis block does not match');
		}
		if (!lastBlockHeader && !matchingGenesisBlock) {
			return false;
		}
		return true;
	}

	public validateGenesisBlock(block: Block): void {
		validateGenesisBlock(block);
	}

	public async verifyBlock(block: Block): Promise<void> {
		block.validate();
		const transactionIDs = [];
		let payloadSize = 0;
		for (const tx of block.payload) {
			transactionIDs.push(tx.id);
			payloadSize += tx.getBytes().length;
		}
		if (payloadSize > this.constants.maxPayloadLength) {
			throw new Error(
				`Payload length is longer than configured length: ${this.constants.maxPayloadLength}.`,
			);
		}
		const tree = new MerkleTree();
		await tree.init(transactionIDs);
		if (!tree.root.equals(block.header.transactionRoot as Buffer)) {
			throw new Error('Invalid transaction root.');
		}
	}

	public async saveBlock(
		block: Block,
		stateStore: StateStore,
		finalizedHeight: number,
		{ removeFromTempTable } = {
			removeFromTempTable: false,
		},
	): Promise<void> {
		const isConsecutiveBlock =
			this.lastBlock.header.height + 1 === block.header.height &&
			block.header.previousBlockID.equals(this.lastBlock.header.id);

		if (!isConsecutiveBlock) {
			throw new Error('Non consecutive block cannot be saved.');
		}
		await this.dataAccess.saveBlock(block, stateStore, finalizedHeight, removeFromTempTable);
		this.dataAccess.addBlockHeader(block.header);
		this._finalizedHeight = finalizedHeight;
		this._lastBlock = block;
	}

	public async removeBlock(
		block: Block,
		stateStore: StateStore,
		{ saveTempBlock } = { saveTempBlock: false },
	): Promise<void> {
		if (block.header.version === GENESIS_BLOCK_VERSION) {
			throw new Error('Cannot delete genesis block');
		}
		let secondLastBlock: Block;
		try {
			secondLastBlock = await this.dataAccess.getBlockByID(block.header.previousBlockID);
		} catch (error) {
			throw new Error('PreviousBlock is null');
		}

		await this.dataAccess.deleteBlock(block, stateStore, saveTempBlock);
		await this.dataAccess.removeBlockHeader(block.header.id);
		this._lastBlock = secondLastBlock;
	}

	private async _cacheBlockHeaders(storageLastBlock: Block): Promise<void> {
		// Cache the block headers (size=DEFAULT_MAX_BLOCK_HEADER_CACHE)
		const fromHeight = Math.max(storageLastBlock.header.height - DEFAULT_MAX_BLOCK_HEADER_CACHE, 0);
		const toHeight = storageLastBlock.header.height;

		debug(
			{ h: storageLastBlock.header.height, fromHeight, toHeight },
			'Cache block headers during chain init',
		);
		const blockHeaders = await this.dataAccess.getBlockHeadersByHeightBetween(fromHeight, toHeight);
		const sortedBlockHeaders = [...blockHeaders].sort(
			(a: BlockHeader, b: BlockHeader) => a.height - b.height,
		);

		for (const blockHeader of sortedBlockHeaders) {
			debug({ height: blockHeader.height }, 'Add block header to cache');
			this.dataAccess.addBlockHeader(blockHeader);
		}
	}
}
