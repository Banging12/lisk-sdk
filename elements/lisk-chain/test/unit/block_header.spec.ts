/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { hash } from '@liskhq/lisk-cryptography';
import { BlockHeader } from '../../src/block_header';
import {
	blockHeaderSchema,
	blockHeaderSchemaWithId,
	signingBlockHeaderSchema,
} from '../../src/schema';

const getBlockAttrs = () => ({
	version: 1,
	timestamp: 1009988,
	height: 1009988,
	previousBlockID: Buffer.from('4a462ea57a8c9f72d866c09770e5ec70cef18727', 'hex'),
	stateRoot: Buffer.from('7f9d96a09a3fd17f3478eb7bef3a8bda00e1238b', 'hex'),
	transactionRoot: Buffer.from('b27ca21f40d44113c2090ca8f05fb706c54e87dd', 'hex'),
	assetsRoot: Buffer.from('b27ca21f40d44113c2090ca8f05fb706c54e87dd', 'hex'),
	generatorAddress: Buffer.from('be63fb1c0426573352556f18b21efd5b6183c39c', 'hex'),
	maxHeightPrevoted: 1000988,
	maxHeightGenerated: 1000988,
	validatorsHash: hash(Buffer.alloc(0)),
	aggregateCommit: {
		height: 0,
		aggregationBits: Buffer.alloc(0),
		certificateSignature: Buffer.alloc(0),
	},
	signature: Buffer.from('6da88e2fd4435e26e02682435f108002ccc3ddd5', 'hex'),
});

const blockId = Buffer.from(
	'097ce5adc1a34680d6c939287011dec9b70a3bc1f5f896a3f9024fc9bed59992',
	'hex',
);

const blockHeaderBytes = Buffer.from(
	'080110c4d23d18c4d23d22144a462ea57a8c9f72d866c09770e5ec70cef187272a14be63fb1c0426573352556f18b21efd5b6183c39c3214b27ca21f40d44113c2090ca8f05fb706c54e87dd3a14b27ca21f40d44113c2090ca8f05fb706c54e87dd42147f9d96a09a3fd17f3478eb7bef3a8bda00e1238b489c8c3d509c8c3d5a20e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8556206080012001a006a146da88e2fd4435e26e02682435f108002ccc3ddd5',
	'hex',
);

const blockHeaderProps = [
	'version',
	'timestamp',
	'height',
	'previousBlockID',
	'generatorAddress',
	'transactionRoot',
	'assetsRoot',
	'stateRoot',
	'maxHeightPrevoted',
	'maxHeightGenerated',
	'validatorsHash',
	'aggregateCommit',
];

describe('block_header', () => {
	describe('signingBlockHeaderSchema', () => {
		it('should be valid schema without signature and id', () => {
			expect(signingBlockHeaderSchema).toMatchSnapshot();
			expect(Object.keys(signingBlockHeaderSchema.properties)).toIncludeAllMembers([
				...blockHeaderProps,
			]);
			expect(Object.keys(signingBlockHeaderSchema.properties)).not.toIncludeAllMembers([
				'id',
				'signature',
			]);
		});
	});

	describe('blockHeaderSchema', () => {
		it('should be valid schema with signature but without id', () => {
			expect(blockHeaderSchema).toMatchSnapshot();
			expect(Object.keys(blockHeaderSchema.properties)).toIncludeAllMembers([
				...blockHeaderProps,
				'signature',
			]);
			expect(Object.keys(blockHeaderSchema.properties)).not.toIncludeAllMembers(['id']);
		});
	});

	describe('blockHeaderSchemaWithId', () => {
		it('should be valid schema with signature and id', () => {
			expect(blockHeaderSchemaWithId).toMatchSnapshot();
			expect(Object.keys(blockHeaderSchemaWithId.properties)).toIncludeAllMembers([
				...blockHeaderProps,
				'signature',
				'id',
			]);
		});
	});

	describe('BlockHeader', () => {
		describe('constructor', () => {
			it('should initialize block header object', () => {
				const data = getBlockAttrs();

				const blockHeader = new BlockHeader(data);

				expect(blockHeader).toBeInstanceOf(BlockHeader);
				expect(blockHeader.version).toEqual(data.version);
				expect(blockHeader.timestamp).toEqual(data.timestamp);
				expect(blockHeader.height).toEqual(data.height);
				expect(blockHeader.generatorAddress).toEqual(data.generatorAddress);
				expect(blockHeader.previousBlockID).toEqual(data.previousBlockID);
				expect(blockHeader.stateRoot).toEqual(data.stateRoot);
				expect(blockHeader.validatorsHash).toEqual(data.validatorsHash);
				expect(blockHeader.aggregateCommit).toEqual(data.aggregateCommit);
				expect(blockHeader.maxHeightPrevoted).toEqual(data.maxHeightPrevoted);
				expect(blockHeader.maxHeightGenerated).toEqual(data.maxHeightGenerated);
				expect(blockHeader.assetsRoot).toEqual(data.assetsRoot);
				expect(blockHeader.transactionRoot).toEqual(data.transactionRoot);
				expect(blockHeader.id).toEqual(blockId);
				expect(blockHeader.signature).toEqual(data.signature);
			});
		});

		describe('fromBytes', () => {
			it('should load the block header from bytes', () => {
				const blockHeader = BlockHeader.fromBytes(blockHeaderBytes);

				expect(blockHeader.toObject()).toEqual({ ...getBlockAttrs(), id: blockId });
			});
		});

		describe('validate', () => {
			it('should validate the block header without error', () => {
				const data = getBlockAttrs();
				const blockHeader = new BlockHeader(data);

				expect(blockHeader.validate()).toBeUndefined();
			});

			it('should throw error if previous block id is not set', () => {
				const data = getBlockAttrs();
				const blockHeader = new BlockHeader({ ...data, previousBlockID: Buffer.alloc(0) });

				expect(() => blockHeader.validate()).toThrow('Previous block id must not be empty.');
			});
		});

		describe('signature', () => {
			it('should throw error if block header not signed', () => {
				const { signature, ...rest } = getBlockAttrs();
				const blockHeader = new BlockHeader(rest);

				expect(() => blockHeader.signature).toThrow('Block header is not signed.');
			});
		});

		describe('id', () => {
			it('should throw error if block header not signed', () => {
				const { signature, ...rest } = getBlockAttrs();
				const blockHeader = new BlockHeader(rest);

				expect(() => blockHeader.id).toThrow('Can not generate the id for unsigned block header.');
			});

			it('should not throw error if block header is signed', () => {
				const data = getBlockAttrs();
				const blockHeader = new BlockHeader(data);

				expect(blockHeader.id).toEqual(blockId);
			});
		});
	});
});
