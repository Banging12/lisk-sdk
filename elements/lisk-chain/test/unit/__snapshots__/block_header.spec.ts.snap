// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`block_header blockHeaderSchema should be valid schema with signature but without id 1`] = `
Object {
  "$id": "/block/header/3",
  "properties": Object {
    "aggregateCommit": Object {
      "fieldNumber": 12,
      "properties": Object {
        "aggregationBits": Object {
          "dataType": "bytes",
          "fieldNumber": 2,
        },
        "certificateSignature": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "height",
        "aggregationBits",
        "certificateSignature",
      ],
      "type": "object",
    },
    "assetsRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "generatorAddress": Object {
      "dataType": "bytes",
      "fieldNumber": 5,
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "maxHeightGenerated": Object {
      "dataType": "uint32",
      "fieldNumber": 10,
    },
    "maxHeightPrevoted": Object {
      "dataType": "uint32",
      "fieldNumber": 9,
    },
    "previousBlockID": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "signature": Object {
      "dataType": "bytes",
      "fieldNumber": 13,
    },
    "stateRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 8,
    },
    "timestamp": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "transactionRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "validatorsHash": Object {
      "dataType": "bytes",
      "fieldNumber": 11,
    },
    "version": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "version",
    "timestamp",
    "height",
    "previousBlockID",
    "generatorAddress",
    "transactionRoot",
    "assetsRoot",
    "stateRoot",
    "maxHeightPrevoted",
    "maxHeightGenerated",
    "validatorsHash",
    "aggregateCommit",
    "signature",
  ],
  "type": "object",
}
`;

exports[`block_header blockHeaderSchemaWithId should be valid schema with signature and id 1`] = `
Object {
  "$id": "/block/header/3",
  "properties": Object {
    "aggregateCommit": Object {
      "fieldNumber": 12,
      "properties": Object {
        "aggregationBits": Object {
          "dataType": "bytes",
          "fieldNumber": 2,
        },
        "certificateSignature": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "height",
        "aggregationBits",
        "certificateSignature",
      ],
      "type": "object",
    },
    "assetsRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "generatorAddress": Object {
      "dataType": "bytes",
      "fieldNumber": 5,
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "id": Object {
      "dataType": "bytes",
      "fieldNumber": 14,
    },
    "maxHeightGenerated": Object {
      "dataType": "uint32",
      "fieldNumber": 10,
    },
    "maxHeightPrevoted": Object {
      "dataType": "uint32",
      "fieldNumber": 9,
    },
    "previousBlockID": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "signature": Object {
      "dataType": "bytes",
      "fieldNumber": 13,
    },
    "stateRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 8,
    },
    "timestamp": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "transactionRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "validatorsHash": Object {
      "dataType": "bytes",
      "fieldNumber": 11,
    },
    "version": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "version",
    "timestamp",
    "height",
    "previousBlockID",
    "generatorAddress",
    "transactionRoot",
    "assetsRoot",
    "stateRoot",
    "maxHeightPrevoted",
    "maxHeightGenerated",
    "validatorsHash",
    "aggregateCommit",
    "signature",
    "id",
  ],
  "type": "object",
}
`;

exports[`block_header signingBlockHeaderSchema should be valid schema without signature and id 1`] = `
Object {
  "$id": "/block/header/signing/3",
  "properties": Object {
    "aggregateCommit": Object {
      "fieldNumber": 12,
      "properties": Object {
        "aggregationBits": Object {
          "dataType": "bytes",
          "fieldNumber": 2,
        },
        "certificateSignature": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "height",
        "aggregationBits",
        "certificateSignature",
      ],
      "type": "object",
    },
    "assetsRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "generatorAddress": Object {
      "dataType": "bytes",
      "fieldNumber": 5,
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "maxHeightGenerated": Object {
      "dataType": "uint32",
      "fieldNumber": 10,
    },
    "maxHeightPrevoted": Object {
      "dataType": "uint32",
      "fieldNumber": 9,
    },
    "previousBlockID": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "stateRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 8,
    },
    "timestamp": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "transactionRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "validatorsHash": Object {
      "dataType": "bytes",
      "fieldNumber": 11,
    },
    "version": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "version",
    "timestamp",
    "height",
    "previousBlockID",
    "generatorAddress",
    "transactionRoot",
    "assetsRoot",
    "stateRoot",
    "maxHeightPrevoted",
    "maxHeightGenerated",
    "validatorsHash",
    "aggregateCommit",
  ],
  "type": "object",
}
`;
